### ISSUE TITLE
URL-Encoded secret/key in `Authorisation` header

### ISSUE THREAD

**Issue by freegroup** [panva/node-openid-client#236]
I'm using your lib with our OpenId-Connect server but unfortunately, we have some issues regarding
the `Authorisation: Basic ` header. You encode the `user:password` with url_encode. 

In the RFC for the Authorisation, it is mention that the user/password is just base64
https://tools.ietf.org/html/rfc7617 like `base64(user+":"+password)` but in your impl it is `base64(url_encode(user)+":"+url_encode(password))`, which breaks passwords with special character

https://github.com/panva/node-openid-client/blob/2dfa313f24bab351344932b7b1c67c9cab6031f8/lib/helpers/client.js#L98-L100

I'm wrong or your code :-)

**Comment by panva** [comment]
Hi @freegroup 

This is correct behaviour, see https://tools.ietf.org/html/rfc6749#section-2.3.1 and the related appendix.

> The client identifier is encoded using the
> "application/x-www-form-urlencoded" encoding algorithm per
> Appendix B, and the encoded value is used as the username; the client
> password is encoded using the same algorithm and used as the
> password.

If the AS you're using openid-client with does not process credentials with special characters correctly then you should request them to fix it as per the specification's appendix. If you can't get them to fix client_secret_basic see if you can use client_secret_post instead or request a client_id/secret without special characters that need URL form encoding

Note this behaviour is certified with OIDC Certification Suite where this behaviour was also missing and was [added](https://github.com/openid-certification/oidctest/issues/88)

Bottom line, Basic Auth !== client_secret_basic.
